[css-scroll-snap] Focusing an element with scroll snap should not always result in snapping to that element

nmoucht opened the pull request:
#### 7c41fd985950f27acd9a3a55c9c6db4d123bd6ae
<pre>
[css-scroll-snap] Focusing an element with scroll snap should not always result in snapping to that element
<a href="https://bugs.webkit.org/show_bug.cgi?id=249354">https://bugs.webkit.org/show_bug.cgi?id=249354</a>
rdar://103731027

Reviewed by Simon Fraser.

This patch refactors the current implementation of this chunk of the spec: &quot;If multiple boxes
were snapped before and their snap positions no longer coincide, then if one of them is
focused or targeted, the scroll container must re-snap to that one and otherwise which one
to re-snap to is UA-defined.&quot; We now more closely follow the language of the spec by specifically
keeping track of the boxes we are currently snapped to through m_currentlySnappedBoxesX/Y.
Vectors are necessary as we can be snapped to multiple boxes per axis (ie. multiple boxes along
the same snap offset). If we see that the number of boxes we are snapped to changes due to
relayout, we use the algorithm the spec specifies by first checking if any of the previously
snapped boxes are focused, then if not, choosing a UA defined box. This also fixes the focus
bug as we are now properly snapping to a focused element only under the correct scenario
(when we were previously snapped to multiple boxes). Currently
css/css-scroll-snap/snap-after-relayout/resnap-to-focused.html is failing due to getting
the scroll position before we finish triggering relayout, so we need to figure out a way
to fix the test.

* Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp:
(WebCore::updateSnapOffsetsForScrollableArea):
(WebCore::convertOffsetInfo):
* Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.h:
* Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
(WebCore::ScrollSnapAnimatorState::setFocusedElementForAxis):
(WebCore::ScrollSnapAnimatorState::preserveCurrentTargetForAxis):
(WebCore::ScrollSnapAnimatorState::currentlySnappedOffsets):
(WebCore::isSnappedToMultipleBoxes):
(WebCore::ScrollSnapAnimatorState::chooseBoxToResnapTo):
(WebCore::ScrollSnapAnimatorState::resnapAfterLayout):
* Source/WebCore/platform/ScrollSnapAnimatorState.h:

Canonical link: <a href="https://commits.webkit.org/259381@main">https://commits.webkit.org/259381@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7e766978f45f8e658c2f71e1c073e7071bc8cdbd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104755 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/13831 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/37671 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/114030 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/174228 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108670 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14963 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4762 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/97086 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/112955 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110515 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/11570 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/37671 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/97086 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitpy~~](https://ews-build.webkit.org/#/builders/5/builds/108218 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/93416 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/37671 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/97086 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/94672 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7187 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/37671 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/92645 "Built successfully") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/83/builds/4932 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/7288 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/4126 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/30201 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 services~~](https://ews-build.webkit.org/#/builders/20/builds/103568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13340 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/37671 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/101334 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6467 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/9074 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/45/builds/25274 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-snap/snap-after-relayout/snap-to-different-targets-expected.txt:
```diff
@@ -1,3 +1,3 @@
 
-PASS Scroller should snap to at least one of the targets if unable to snap toboth after a layout change.
+FAIL Scroller should snap to at least one of the targets if unable to snap toboth after a layout change. assert_true: expected true got false
```
Is this FAIL expected?

nmoucht replied:
I mention this at the end of the changelog but this is another case of getting the scroll position before we run our relayout code. I will create a follow up bug to fix this test.

smfr commented on Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp:
```diff
@@ -324,6 +324,7 @@ void updateSnapOffsetsForScrollableArea(ScrollableArea& scrollableArea, const Re
     HashMap<LayoutUnit, SnapOffset<LayoutUnit>> verticalSnapOffsetsMap;
     HashMap<LayoutUnit, SnapOffset<LayoutUnit>> horizontalSnapOffsetsMap;
     Vector<LayoutRect> snapAreas;
+    Vector<uint64_t> snapAreasIDs;
```
Using bare `uint64_t` is discouraged; use an `ObjectIdentifier<>` instead.

nmoucht replied:
Change to ElementIdentifier alias.

smfr commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -95,49 +95,73 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
 }
 
-void ScrollSnapAnimatorState::setFocusedElementForAxis(ScrollEventAxis axis)
+bool ScrollSnapAnimatorState::preserveCurrentTargetForAxis(ScrollEventAxis axis, uint64_t box)
```
`box` is a `boxIdentifier`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -95,49 +95,73 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
 }
 
-bool ScrollSnapAnimatorState::resnapAfterLayout(ScrollOffset scrollOffset, const ScrollExtents& scrollExtents, float pageScale)
+Vector<SnapOffset<LayoutUnit>> ScrollSnapAnimatorState::currentlySnappedOffsets(ScrollEventAxis axis)
```
Maybe call this `currentlySnappedOffsetsForAxis`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -95,49 +95,73 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
+    return currentlySnappedOffsets;
+}
+
+static bool isSnappedToMultipleBoxes(Vector<SnapOffset<LayoutUnit>> verticalOffsets, Vector<SnapOffset<LayoutUnit>> horizontalOffsets)
```
Arguments should be const references.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -95,49 +95,73 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
+static bool isSnappedToMultipleBoxes(Vector<SnapOffset<LayoutUnit>> verticalOffsets, Vector<SnapOffset<LayoutUnit>> horizontalOffsets)
+{
+    HashSet<uint64_t> boxIDs;
+    verticalOffsets.appendVector(horizontalOffsets);
```
This is confusing; it's OK to append the vectors because all you care about is the snapTargetIDs? Seems like there would be a better way to write this code.

nmoucht replied:
It's a bit tricky since we are tracking the offsets per axis but we want to perform an operation on all of them. I have changed this and the other sights to explicitly make a separate vector which should hopefully make this clearer.

smfr commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -95,49 +95,73 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
+    return boxIDs.size() > 1 || hasMultipleSnapAreas;
+}
+
+uint64_t ScrollSnapAnimatorState::chooseBoxToResnapTo(Vector<SnapOffset<LayoutUnit>> verticalOffsets, Vector<SnapOffset<LayoutUnit>> horizontalOffsets)
```
Arguments should be const references. Return an ObjectIdentifier<>.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -95,49 +95,73 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
+{
+    HashSet<uint64_t> boxIDs;
+    
+    verticalOffsets.appendVector(horizontalOffsets);
```
Confusing again.

nmoucht replied:
See above.

smfr commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -95,49 +95,73 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
+    }
+    auto snapOffsets = snapOffsetsForAxis(ScrollEventAxis::Horizontal);
+    
+    snapOffsets.appendVector(snapOffsetsForAxis(ScrollEventAxis::Vertical));
```
More confusing mixing of X and Y offsets.

nmoucht replied:
See above.

smfr commented on Source/WebCore/platform/ScrollSnapAnimatorState.h:
```diff
@@ -111,8 +98,10 @@ class ScrollSnapAnimatorState {
-    void setFocusedElementForAxis(ScrollEventAxis);
+    bool preserveCurrentTargetForAxis(ScrollEventAxis, uint64_t);
+    bool setFocusedElementForAxis(ScrollEventAxis);
+    Vector<SnapOffset<LayoutUnit>> currentlySnappedOffsets(ScrollEventAxis);
```
Function should be const.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/ScrollSnapAnimatorState.h:
```diff
@@ -111,8 +98,10 @@ class ScrollSnapAnimatorState {
     void transitionToDestinationReachedState();
-    bool preserveCurrentTargetForAxis(ScrollEventAxis);
-    void setFocusedElementForAxis(ScrollEventAxis);
+    bool preserveCurrentTargetForAxis(ScrollEventAxis, uint64_t);
```
What does the return value mean? Use an enum or comment.

nmoucht replied:
Added comment to explain.

smfr commented on Source/WebCore/platform/ScrollSnapAnimatorState.h:
```diff
@@ -111,8 +98,10 @@ class ScrollSnapAnimatorState {
-    bool preserveCurrentTargetForAxis(ScrollEventAxis);
-    void setFocusedElementForAxis(ScrollEventAxis);
+    bool preserveCurrentTargetForAxis(ScrollEventAxis, uint64_t);
+    bool setFocusedElementForAxis(ScrollEventAxis);
```
What does the return value mean? Use an enum or comment.

nmoucht replied:
Removed this function.

nt1m commented on Source/WebCore/platform/ScrollSnapAnimatorState.h:
```diff
@@ -111,8 +98,9 @@ class ScrollSnapAnimatorState {
-    bool preserveCurrentTargetForAxis(ScrollEventAxis);
-    void setFocusedElementForAxis(ScrollEventAxis);
+    bool preserveCurrentTargetForAxis(ScrollEventAxis, ElementIdentifier);
+    Vector<SnapOffset<LayoutUnit>> currentlySnappedOffsetsForAxis(ScrollEventAxis axis) const;
```
```suggestion
    Vector<SnapOffset<LayoutUnit>> currentlySnappedOffsetsForAxis(ScrollEventAxis) const;
```

to fix style bot

nmoucht replied:
Fixed.

smfr commented on LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-snap/snap-after-relayout/focus-element-no-snap.html:
```diff
@@ -0,0 +1,52 @@
+#snapper {
+    width: 100px;
+    height: 200px;
+    /* margin: 15vh auto; */
```
Remove

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -95,49 +95,81 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
+        boxIDs.add(offset.snapTargetID);
+        hasMultipleSnapAreas |= offset.snapAreaIndices.size() > 1;
+    }
+    return boxIDs.size() > 1 || hasMultipleSnapAreas;
```
I think this would be more efficient if you just looped over the X and Y offsets in two loops, and early returned as soon as you find the exit condition.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -95,49 +95,81 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
+    return boxIDs.size() > 1 || hasMultipleSnapAreas;
+}
+
+ElementIdentifier ScrollSnapAnimatorState::chooseBoxToResnapTo(const Vector<SnapOffset<LayoutUnit>>& verticalOffsets, const Vector<SnapOffset<LayoutUnit>>& horizontalOffsets)
```
I think this function can be `const`. Normally we put horizontal things first.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -95,49 +95,81 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
+{
+    HashSet<ElementIdentifier> boxIDs;
+    
+    Vector<SnapOffset<LayoutUnit>> currentlySnappedOffsets;
```
You don't use this at all.

nmoucht replied:
Changed to use two loops.

smfr commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -95,49 +95,81 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
+
+ElementIdentifier ScrollSnapAnimatorState::chooseBoxToResnapTo(const Vector<SnapOffset<LayoutUnit>>& verticalOffsets, const Vector<SnapOffset<LayoutUnit>>& horizontalOffsets)
+{
+    HashSet<ElementIdentifier> boxIDs;
```
Maybe snappedBoxIDs.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -95,49 +95,81 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
+    }
+    
+    Vector<SnapOffset<LayoutUnit>> possibleOffsets;
+    possibleOffsets.appendVector(snapOffsetsForAxis(ScrollEventAxis::Horizontal));
```
I was confused by the relationship between the arguments (e.g. `horizontalOffsets `) and what `snapOffsetsForAxis(ScrollEventAxis::Horizontal)` gives you. I think the argument is "current" and "snapOffsetsForAxis()" is computing the new state? It would be clearer to call `snapOffsetsForAxis()` in the caller, and explicitly pass the new ones. This function can access `m_currentlySnappedBoxesX` and  `m_currentlySnappedBoxesY` for the current state, right?

smfr replied:
You could even make a helper function that just returns the `HashSet<ElementIdentifier>` of currently snapped boxIDs.

nmoucht replied:
Ah I see yes the idea is comparing the currently snapped boxes (`m_currentlySnappedBoxesX` and `m_currentlySnappedBoxesY`) and the newly snapped boxes (snapOffsetsForAxis). You are right that it makes more sense to pass the newly snapped boxes and access the member variables from this function. Also added the suggested helper function `currentlySnappedBoxes()`.

smfr commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -95,49 +95,81 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
+    possibleOffsets.appendVector(snapOffsetsForAxis(ScrollEventAxis::Horizontal));
+    possibleOffsets.appendVector(snapOffsetsForAxis(ScrollEventAxis::Vertical));
+    
+    auto found = std::find_if(possibleOffsets.begin(), possibleOffsets.end(), [boxIDs](SnapOffset<LayoutUnit> p) -> bool {
```
Two loops here to avoid the vector copying would be fine.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -151,23 +183,19 @@ bool ScrollSnapAnimatorState::resnapAfterLayout(ScrollOffset scrollOffset, const
+    auto horizontalOffsets = currentlySnappedOffsetsForAxis(ScrollEventAxis::Horizontal);
     
-    LOG_WITH_STREAM(ScrollSnap, stream << "ScrollSnapAnimatorState::resnapAfterLayout() current target: horizontal: " << activeHorizontalID << " vertical: "<< activeVerticalID);
+    if (isSnappedToMultipleBoxes(m_currentlySnappedBoxesX, m_currentlySnappedBoxesY) && !isSnappedToMultipleBoxes(verticalOffsets, horizontalOffsets)) {
```
You should keep the logging and have it indicate what the result of resnapping was.

nmoucht replied:
Fixed.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-snap/snap-after-relayout/focus-element-no-snap.html:
```diff
@@ -0,0 +1,52 @@
+<!doctype html>
+<title>Resnap to focused element after relayout</title>
```
Can you export this to WPT? `export-w3c-test-changes -b 249354 -g HEAD -c --no-linter`

Curious to see cross browser results for this test.

nmoucht replied:
https://github.com/web-platform-tests/wpt/pull/38151

nt1m replied:
Seems to pass in other browsers 👍 

A bot should approve it when your change lands in WebKit.

nt1m commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -95,49 +95,94 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
+}
+
+static bool isSnappedToMultipleBoxes(const Vector<SnapOffset<LayoutUnit>>& verticalOffsets, const Vector<SnapOffset<LayoutUnit>>& horizontalOffsets)
+{
```
Should it be horizontal first?

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -95,49 +95,94 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
+            return true;
+    }
+    
+    for (auto offset : horizontalOffsets) {
```
Do horizontal first (unless it matters the vertical takes priority somehow).

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -95,49 +95,94 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
+{
+    HashSet<ElementIdentifier> boxIDs;
+    
+    for (auto offset : verticalOffsets) {
```
`const auto& offset` perhaps to avoid the copy.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -95,49 +95,94 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
+    return snappedBoxIDs;
+}
+
+ElementIdentifier ScrollSnapAnimatorState::chooseBoxToResnapTo(const Vector<SnapOffset<LayoutUnit>>& horizontalOffsets, const Vector<SnapOffset<LayoutUnit>>& verticalOffsets) const
```
This is much cleaner now!

smfr commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -151,23 +196,20 @@ bool ScrollSnapAnimatorState::resnapAfterLayout(ScrollOffset scrollOffset, const
-        }
-    }
+    auto verticalOffsets = currentlySnappedOffsetsForAxis(ScrollEventAxis::Vertical);
+    auto horizontalOffsets = currentlySnappedOffsetsForAxis(ScrollEventAxis::Horizontal);
```
horizontal first for consistency.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -151,23 +196,20 @@ bool ScrollSnapAnimatorState::resnapAfterLayout(ScrollOffset scrollOffset, const
+        snapPointChanged |= preserveCurrentTargetForAxis(ScrollEventAxis::Vertical, box);
+        LOG_WITH_STREAM(ScrollSnap, stream << "ScrollSnapAnimatorState::resnapAfterLayout() current target: " << box << " snapPointChanged: " << snapPointChanged);
+    }
+    if (isSnappedToMultipleBoxes(m_currentlySnappedBoxesX, m_currentlySnappedBoxesY) != isSnappedToMultipleBoxes(verticalOffsets, horizontalOffsets)) {
```
Should we cache the results of `isSnappedToMultipleBoxes` to avoid calling it twice a second time?

nmoucht replied:
Fixed.

nmoucht committed:
[css-scroll-snap] Focusing an element with scroll snap should not always result in snapping to that element
https://bugs.webkit.org/show_bug.cgi?id=249354
rdar://103731027

Reviewed by Simon Fraser.

This patch refactors the current implementation of this chunk of the spec: "If multiple boxes
were snapped before and their snap positions no longer coincide, then if one of them is
focused or targeted, the scroll container must re-snap to that one and otherwise which one
to re-snap to is UA-defined." We now more closely follow the language of the spec by specifically
keeping track of the boxes we are currently snapped to through m_currentlySnappedBoxesX/Y.
Vectors are necessary as we can be snapped to multiple boxes per axis (ie. multiple boxes along
the same snap offset). If we see that the number of boxes we are snapped to changes due to
relayout, we use the algorithm the spec specifies by first checking if any of the previously
snapped boxes are focused, then if not, choosing a UA defined box. This also fixes the focus
bug as we are now properly snapping to a focused element only under the correct scenario
(when we were previously snapped to multiple boxes). Currently
css/css-scroll-snap/snap-after-relayout/resnap-to-focused.html is failing due to getting
the scroll position before we finish triggering relayout, so we need to figure out a way
to fix the test.

* Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp:
(WebCore::updateSnapOffsetsForScrollableArea):
(WebCore::convertOffsetInfo):
* Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.h:
* Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
(WebCore::ScrollSnapAnimatorState::setFocusedElementForAxis):
(WebCore::ScrollSnapAnimatorState::preserveCurrentTargetForAxis):
(WebCore::ScrollSnapAnimatorState::currentlySnappedOffsets):
(WebCore::isSnappedToMultipleBoxes):
(WebCore::ScrollSnapAnimatorState::chooseBoxToResnapTo):
(WebCore::ScrollSnapAnimatorState::resnapAfterLayout):
* Source/WebCore/platform/ScrollSnapAnimatorState.h:

Canonical link: https://commits.webkit.org/259381@main
